Remove CSVConfig and use CSV Generator for configurable inputs and output

hasbro17 opened the pull request:
**Description of the change:**
Continuation of #2407 to implement the CSV generator.

> - cmd/operator-sdk/olmcatalog: remove config flag, add include flag, use CSV and package manifest generators
> 
> - internal/generate/olm-catalog: implement CSV generator, moving most of the code from internal/scaffold/olm-catalog verbatim. Also removed CSV config
> 
> - doc: remove CSVConfig documentation, update CLI docs

Currently this gets #2407 rebased with the master and breaks up the `internal/generate/olm-catalog` pkg changes, and the `internal/generate/testdata/` manifests changes into separate commits for ease of reviewing.

**TODO:**
- Change `--include` flag to `--inputs` to allow configuring the CSV generator `Inputs` field for Deploy and API directory paths  e.g: 
    `--input deploy=./deploy/production/...,apis=./pkg/apis`
- Remove `Filters` field from generator config since its UX was ambiguous
- Add flag `--outputDir` to set CSV Generator `Output` field to set the output root directory for the `olm-catalog` bundles directory.

**Motivation for the change:**

Resolves #2493

hasbro17 commented:
@estroz No need to review until I finish the TODOs but a quick sanity check over the first commit would be helpful to ensure I've carried forward the necessary changes from #2407 

Had to untangle that PR a bit to rebase and split up the commits since quite a few changes from that had already gone into master.

estroz commented on internal/generate/crd/crd_test.go:
```diff
@@ -151,7 +151,7 @@ func TestCRDNonGo(t *testing.T) {
 			cfg := gen.Config{
 				Inputs: map[string]string{
-					CRDsDirKey: c.crdsDir,
+					CRDsDirKey: c.crdsDir
```
This won't compile.

hasbro17 commented:
Yeah I'll remove the Helm tests from this. I initially split it up as a separate commit but a follow up would be easier to review.

hasbro17 commented:
@estroz I've done some manual tests and have this mostly working. 
I still need to clean up a couple of things plus update the unit tests, but I've got a couple of questions that came up as I was refactoring.
I've marked them as TODO comments if you could take a look at those. Just the [2nd commit](https://github.com/operator-framework/operator-sdk/pull/2511/commits/d2eea3ddf6902c1eb2c45e7561eaa38f36046199) which has my changes to your old PR.

hasbro17 commented on internal/generate/olm-catalog/csv.go:
```diff
@@ -44,13 +44,20 @@ import (
-	BundleDirKey = "bundle"
+	// Input keys for CSV generator whose values are the filepaths for the respective input directories
+
+	// DEBUG: Why is this an input key? Users don't need to configure this.
```
@estroz

hasbro17 replied:
Figured this out as I'm writing the test for this. 
`BundleDirKey` is just being used to check for a preexisting CSV bundle directory so we can update from that when generating the CSV.

Going to refactor `BundleDirKey ` to not be a generator key and just be a private field of the  `csvGenerator` since it's something internal and not meant to be an input.

hasbro17 commented on internal/generate/olm-catalog/package_manifest.go:
```diff
@@ -26,16 +26,18 @@ import (
+	packageManifestFileExt = ".package.yaml"
 
-	ManifestsDirKey = "manifests"
+	// TODO: Is this an input that needs to be set via the csv generator inputs flag?
```
@estroz

hasbro17 replied:
Looks like it's the same thing here as with `BundleDirKey `.
Not really a generator input that ever needs to be set. 
Only place this is actually used is in `package_manifest_test.go` to have the generator point to  an existing package manifest data directory.
```Go
inputDir := filepath.Join(testGoDataDir, OLMCatalogDir, testProjectName)
cfg := gen.Config{
	OperatorName: testProjectName,
	Inputs:       map[string]string{manifestsDirKey: inputDir},
}
``` 
Going to replace this with a private field in `pkgGenerator`.

hasbro17 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -124,19 +132,19 @@ func (c csvCmd) run() error {
-		crdManifestSet, err := findCRDs(c.includePaths...)
+		// TODO: Any requirement to update CRD's from any place other
+		// than the input "deploy" directory?
+		// Can CRD manifests be present outside of the "deploy" directory?
```
@estroz

hasbro17 replied:
> They will be present in the filepath.Join(c.inputs[gencatalog.DeployDirKey], "crds") dir, and not elsewhere

Since we're doing a `filepath.Walk` in `GetCRDManifestPaths()` it looks like the CRD manifests just need to be present under `c.inputs[gencatalog.DeployDirKey]` i.e `"deploy/..."` or under any subdirectory in that `filepath.Join(c.inputs[gencatalog.DeployDirKey], "foobar")` i.e `"deploy/foobar/..."`

But you're right, we should allow sourcing CRDs for CSV generation from any path.

Do we need a direct input (`--input crds=production/my-app-crds`) for this now to keep the functionality from CSV config file? 
Or is this something we can add in later on and I can put a TODO for this?

estroz replied:
We can add it in a follow-up.

hasbro17 commented on doc/cli/operator-sdk_generate_csv.md:
```diff
@@ -21,12 +19,13 @@ operator-sdk generate csv [flags]
       --default-channel        Use the channel passed to --csv-channel as the package manifests' default channel. Only valid when --csv-channel is set
       --from-version string    Semantic version of an existing CSV to use as a base
   -h, --help                   help for csv
+      --include strings        Paths to include in CSV generation, ex. "deploy/prod,deploy/test". If this flag is set and you want to enable default behavior, you must include "deploy/" in the argument list (default [deploy])
```
Need to regenerate the CLI docs as well.

hasbro17 commented on CHANGELOG.md:
```diff
@@ -22,6 +23,8 @@
 
 ### Removed
 
+- **Breaking Change:** Removed CSV configuration file support (defaulting to deploy/olm-catalog/csv-config.yaml) in favor of including files as input to the generator using [`generate csv --include`](doc/cli/operator-sdk_generate_csv.md#options), defaulting to the `deploy/` directory. ([#2249](https://github.com/operator-framework/operator-sdk/pull/2249))
```
`--include`==>`--inputs` and `--ouputDir`.
Here and above.

hasbro17 commented:
@estroz That should be most of the refactoring done in the last two commits.
I've tested it manually so this should be good for a round of review if you want to take a look at the last two commits.

I'm working on more proper unit test cases specifically for configuring inputs and outputDir for the CSV and package manifest generators but I'll add that in a  separate commit.

estroz commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -97,60 +110,42 @@ Configure CSV generation by writing a config file 'deploy/olm-catalog/csv-config
-	gcfg := gen.Config{
-		OperatorName: c.operatorName,
-		OutputDir:    filepath.Join(gencatalog.OLMCatalogDir, c.operatorName),
+	csv := gencatalog.NewCSV(cfg, c.csvVersion, c.fromVersion)
```
I'm now thinking of using a generator type-specific `Options` struct, ex. `CSVOptions` to configure the generator constructor, and removing `Inputs` and `OutputDir` from `Config`. It would look something like:

```go
type CSVOptions struct {
	gen.Config
	OutputDir 	string
	CSVVersion  string
	FromVersion string
	DeployDir	string
	APIsDir		string
}

func NewCSV(opts CSVOptions) Generator {
	...
}
```

An options struct is idiomatic and makes it easier to deal with defaults. Setup is a little more verbose because you're not only sharing one config, with the tradeoffs of generator-specific configuration captured in one place (no constructor args + inputs map) and the nice properties of struct fields vs those of map indexing.

`OutputDir` may be different for two generators run in the same command, so it makes sense to move it out of `gen.Config` and into generator-specific options.

Using keyed inputs still makes sense; we'd move the key definitions to `cmd/operator-sdk/generate/csv.go`.

hasbro17 replied:
@estroz Agreed. I'd probably go even further to remove the Config altogether if there isn't much else besides the Inputs and OutputDir.
The Inputs and OutputDir in the generator config would be better suited as defined options fields per generator rather than a generic inputs/output for the sake of sharing it between an admittedly small number of generators. And the map indexing is harder to read vs simple option fields.

However to avoid increasing the scope of this PR (and refactoring the CRD generator as well) I'd suggest using the Generator config as it is right now so we can get `generate csv` to have configurable inputs and output flags.

And I'll add a TODO (for a later PR) in the Generator Config to refactor out Inputs and OutputDir into generator specific options.

WDYT?

hasbro17 commented on internal/generate/olm-catalog/csv.go:
```diff
@@ -15,15 +15,436 @@
+	bundleDir := filepath.Join(parentDir, version)
+	_, err := os.Stat(bundleDir)
+	if !os.IsNotExist(err) {
+		// TODO: return and handle this error
```
> !os.IsNotExist(err) == os.IsExist(err) always

I don't think we can use these two checks interchangeably which is why I refactored this a bit. It depends on the kind of error you expect.

If a file already exist then `os.Stat(pathToExistingFile)` will not return an error of the type `os.IsExist(err)`. It will either be `err == nil`(if the file exists) or some other error(permissions, filesystem error etc).

So in `return err == nil || os.IsExist(err)`  `os.IsExist(err)` will always be false for `os.Stat()` so there's no reason to check it.

I can make this a bit more readable however with the following:
```Go
bundleDir := filepath.Join(parentDir, version)
_, err := os.Stat(bundleDir)
if os.IsNotExist(err) { // File does not exist
	return false
} else if err != nil { // some other more serious error
	// TODO: return and handle this error
	log.Fatalf("Failed to stat existing bundle directory %s: %v", bundleDir, err)
}
// File exists, with no error reported by os.Stat()
return true
```

estroz replied:
You are indeed [correct](https://gist.github.com/mastef/05f46d3ab2f5ed6a6787). This block should then be:
```go

func isBundleDirExist(parentDir, version string) (bool, error) {
	// Ensure full path is constructed.
	if parentDir == "" || version == "" {
		return false, nil
	}
	bundleDir := filepath.Join(parentDir, version)
	_, err := os.Stat(bundleDir)
	if err == nil || os.IsExist(err) {
		return true, nil
	}
	if !os.IsNotExist(err) {
		err = fmt.Errorf("error checking if bundle dir %q exists: %v", bundleDir, err)
	}
	return false, err
}
```

and be called in `Generate()` so the error can be returned.

hasbro17 commented on CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@
 - Added support for relative Ansible roles and playbooks paths in the Ansible operator's file. ([#2273](https://github.com/operator-framework/operator-sdk/pull/2273))
 - Add Prometheus metrics support to Ansible-based operators. ([#2179](https://github.com/operator-framework/operator-sdk/pull/2179))
 - On `generate csv`, populate a CSV manifest’s `spec.icon`, `spec.keywords`, and `spec.mantainers` fields with empty values to better inform users how to add data. ([#2521](https://github.com/operator-framework/operator-sdk/pull/2521))
+- Added the [`generate csv --include`](doc/cli/operator-sdk_generate_csv.md#options) option to include files as input to the CSV generator in lieu of a config. ([#2249](https://github.com/operator-framework/operator-sdk/pull/2249))
```
Good catch. Could have sworn I got that line out specifically during the rebase.

hasbro17 commented on internal/generate/olm-catalog/package_manifest.go:
```diff
@@ -26,16 +26,15 @@ import (
-	PackageManifestFileExt = ".package.yaml"
-
-	ManifestsDirKey = "manifests"
+	packageManifestFileExt = ".package.yaml"
```
I'll see but that might be difficult I think.
Since we run both CSV and package manifest generators as part of `generate csv` cmd I need to make sure both of them work correctly per the `--inputs` and `--output-dir` flag changes.

hasbro17 commented on internal/generate/olm-catalog/csv_updaters.go:
```diff
@@ -229,20 +231,64 @@ var _ csvUpdater = crds{}
 // a type if an annotation is present on that type's declaration.
 func (us crds) apply(csv *olmapiv1alpha1.ClusterServiceVersion) error {
-	ownedCRDs := []olmapiv1alpha1.CRDDescription{}
+	ownedDescs := []olmapiv1alpha1.CRDDescription{}
```
Will revert this back to the current [`internal/scaffold/olm-catalog/csv_updaters.go`](https://github.com/operator-framework/operator-sdk/blob/master/internal/scaffold/olm-catalog/csv_updaters.go) if it really is isolated from the rest of the generator changes.

hasbro17 replied:
And add a TODO to move this to `internal/generate/olm-catalog/csv_updaters.go` later.

hasbro17 replied:
@estroz Gave it a shot but that just causes more refactoring since if I move back csv_descriptors.go back to `internal/scaffold/olm-catalog` (and the descriptors pkg as well) then I have to refactor and expose all the structs like [`type crs [][]byte`](https://github.com/operator-framework/operator-sdk/blob/master/internal/scaffold/olm-catalog/csv_updaters.go#L267) in csv_decsriptors to be public since now they're being used by `internal/generate/olm-catalog/csv.go`.

So either I move the entire pkg or not at all. So these changes will have to stay I think.

hasbro17 commented:
@estroz Some updates as requested and other comments. PTAL.
Still working on the new unit tests.

hasbro17 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -53,9 +53,8 @@ for the operator. This file is used to publish the operator to the OLM Catalog.
-Configure CSV generation by writing a config file 'deploy/olm-catalog/csv-config.yaml`,
 		RunE: func(cmd *cobra.Command, args []string) error {
 			// The CSV generator assumes that the deploy and pkg directories are
 			// present at runtime, so this command must be run in a project's root.
```
(Can't comment on the line just below this)
@estroz This isn't true anymore since the inputs and output dir are configurable.
So I'm going to remove the requirement for running in project root i.e check `build/Dockerfile` exists.
I'll need this for the CSV generator to work on kubebuilder's layout plus other users can have non standard layouts.

hasbro17 commented:
@estroz Ready for a full review. 

This last commit adds tests for both the CSV and Package Manifests generators to generate files from a non-standard project layout with Inputs and OutputDir specified.
These tests should also be more complete in that they test the full path for each generator by calling `g.Generate()`

Also I've removed the requirement of `projutil.MustInProjectRoot()` for the cmd as discussed above.

hasbro17 commented:
@aliok I think we have most of the design changes sorted out.
It would be really helpful if you could give this a try for your use case with the knative eventing operator.

Note that we've changed `--include` to `--inputs` where you can pass in the path to your "deploy or operator manifests" directory. 
You can also specify the path for the apis directory as well if needed.

See the CLI docs for this change:
https://github.com/operator-framework/operator-sdk/blob/459e9db87f96e83d7937fa088f01e1164b92793b/doc/cli/operator-sdk_generate_csv.md

Also note, `--output-dir` is back except now it's only the root directory where the entire bundle tree `olm-catalog/...` will be scaffolded. The default is `deploy/`.

So from your previous example this should be something along the lines of:
```
GO111MODULE="on" operator-sdk generate csv            \
  --inputs       "deploy=config" \
  --from-version  "0.11.0"       \
  --csv-version   "0.12.0"       \
  --operator-name knative-eventing-operator \
  --csv-channel   alpha                     \
  --default-channel                         \
  --verbose
```

And the generator won't check for `build/Dockerfile` anymore in the project directory.

Let us know if you spot any issues.

aliok commented:
@hasbro17 

Just gave it a try. Looks like the command is not really using the previous CSV file.

CSV for 0.11.0: https://gist.github.com/aliok/47a5c393e034395e5981aeb7c1a0d979
CSV for 0.12.0, generated by the command: https://gist.github.com/aliok/2d954f480e1952638cf732072f6fa0db

1. Pretty much everything that should come from the previous CSV doesn't exist in the new CSV file.
2. I see lots of files are skipped in the logs.

These make me wonder if I am doing something wrong.

Let me know what I can do to help resolving this issue.

Just in case you needed, here's how I do my tests:

- Checkout https://github.com/aliok/eventing-operator/tree/operator-sdk-tests-for-hasbro
- Execute the command
- Compare with 0.11.0 CSV file in that branch
- Compare with proper 0.12.0 file (in [master](https://github.com/knative/eventing-operator/blob/master/deploy/olm-catalog/knative-eventing-operator/0.12.0/knative-eventing-operator.v0.12.0.clusterserviceversion.yaml))


Command
```
GO111MODULE="on" operator-sdk generate csv            \
  --inputs        "deploy=config,apis=pkg/apis" \
  --from-version  "0.11.0"       \
  --csv-version   "0.12.0"       \
  --operator-name knative-eventing-operator \
  --csv-channel   alpha                     \
  --default-channel                         \
  --update-crds                             \
  --verbose
```

Logs:
```
DEBU[0000] Debug logging is set                         
INFO[0000] Generating CSV manifest version 0.12.0       
INFO[0000] Skipping manifest /v1, Kind=ConfigMap        
WARN[0000] No WATCH_NAMESPACE environment variable nor reference to "metadata.annotations['olm.targetNamespaces']" detected in operator Deployment. For OLM compatibility, your operator MUST watch namespaces defined in "metadata.annotations['olm.targetNamespaces']" 
INFO[0000] Skipping manifest rbac.authorization.k8s.io/v1, Kind=RoleBinding 
INFO[0000] Skipping manifest rbac.authorization.k8s.io/v1, Kind=ClusterRoleBinding 
INFO[0000] Skipping manifest /v1, Kind=ServiceAccount   
INFO[0000] Directory for API operator.knative.dev/v1alpha1, Kind=Eventing does not exist. Skipping CSV annotation parsing for API. 
INFO[0000] Directory for API operator.knative.dev/v1alpha1, Kind=KnativeEventing does not exist. Skipping CSV annotation parsing for API. 
INFO[0000] Fill in the following required fields in file knative-eventing-operator.v0.12.0.clusterserviceversion.yaml:
        spec.description
        spec.keywords
        spec.provider 
DEBU[0000] CSV generator writing deploy/olm-catalog/knative-eventing-operator/0.12.0/knative-eventing-operator.v0.12.0.clusterserviceversion.yaml 
DEBU[0000] Package manifest generator writing deploy/olm-catalog/knative-eventing-operator/knative-eventing-operator.package.yaml 
DEBU[0000] Skipping non CRD manifest config/config-logging.yaml 
DEBU[0000] Skipping non CRD manifest config/config-observability.yaml 
DEBU[0000] Skipping non CRD manifest config/operator.yaml 
DEBU[0000] Skipping non CRD manifest config/role.yaml   
DEBU[0000] Skipping non CRD manifest config/role_binding.yaml 
DEBU[0000] Skipping non CRD manifest config/service_account.yaml 
```

estroz commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -191,25 +182,28 @@ func validateVersion(version string) error {
+		return nil, err
+	}
+	for _, crdPath := range crdPaths {
+		b, err := ioutil.ReadFile(crdPath)
```
If we're not including a `crd` input key in this PR, we should ignore any CRD's coming from `<deploy dir>/olm-catalog`, since they can have the same name as one in `<deploy dir>/crds` with different content.

hasbro17 replied:
Hmm, I think I might just add a `crd` input key while we're here if it's not a big change to remove the ambiguity.
And to guard against re-reading CRDs in any case, I'm going to explicitly ignore `olm-catalog/...` in the `filepath.Walk()` since that's always a constant no matter the inputs/outputDir.

hasbro17 replied:
Added a `--crd-dir` flag and the filepath Walks for CRD manifests now ignore the olm directory.

estroz commented on internal/util/k8sutil/crd.go:
```diff
@@ -62,13 +63,18 @@ func GetCRDManifestPaths(crdsDir string) (crdPaths []string, err error) {
 			typeMeta, err := GetTypeMetaFromBytes(b)
 			if err != nil {
-				return fmt.Errorf("error getting kind from manifest %s: %v", path, err)
+				log.Debugf("Skipping non-manifest file %s: %v", path, err)
```
Are you adding these debug statements for testing or with the intent of permanence?

hasbro17 replied:
These ones are intentional. These will only show up when `generate csv` is run with `--verbose` mode. And in that case it can be useful to see what files are being skipped over for debugging crd updates rather than skipping silently.

Although the goal is to have useful debug statements to surface more context and not just pollute the debug logs so I'm open to rewording or removing them if you see fit.

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -53,14 +53,9 @@ for the operator. This file is used to publish the operator to the OLM Catalog.
 		RunE: func(cmd *cobra.Command, args []string) error {
-			// The CSV generator assumes that the deploy and pkg directories are
-			// present at runtime, so this command must be run in a project's root.
-			projutil.MustInProjectRoot()
```
Why it was removed? Is not required be in the Project root path to gen the CSV in the default path?

hasbro17 replied:
This function does not really verify project root. All it does is check for `build/Dockerfile`, which is overly restrictive for no clear reason, when it comes to running this in non-SDK project layouts like kubebuilder, or SDK projects that move their `build/Dockerfile` to some other path.

And the purpose of this PR is to allow running `generate csv` in non-SDK layouts as much as possible.

camilamacedo86 replied:
Hi @hasbro17, 

Currently, we check if is the root of the project by checking if the Dockerfile exist. I agree that after the integration it will change. We will probably change the implementation of this func to check if has the Project file or not in the current dir as it is done in kubebuilder. However, IMO we cannot remove this now. Note that:

- The command should return the error if it not be executed from the root dir. 
- The command should still work successfully whiteout the need to inform the "inputs". 

Then, we cannot the command be executed outside of the root project.

hasbro17 replied:
@camilamacedo86 This change is not just for after the kubebuilder integration.
It's meant facilitate  running the `generate csv` command in a non-SDK project today.
E.g: https://github.com/knative/eventing-operator


> The command should return the error if it not be executed from the root dir

Why should the command not be run from the root directory?

> The command should still work successfully whiteout the need to inform the "inputs"

It will still do that without this check if you run it without any `inputs` flag in the root directory.

> we cannot the command be executed outside of the root project

I don't see any reason why not. The default still works. We're making the exception easier here i.e to run in a project that doesn't have `build/Dockerfile`. That's all our definition of root directory is anyway.

camilamacedo86 replied:
What you said make completed sense for me. 👍 +1
I think you are 1000% right here.

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -74,14 +69,28 @@ Configure CSV generation by writing a config file 'deploy/olm-catalog/csv-config
+		},
+		`Key value input paths used in CSV generation.
+Use this to set custom paths for operator manifests and API type definitions
+E.g: --inputs deploy=config/production,apis=pkg/myapp/apis 
```
WDYT about use the examples as [here](https://github.com/operator-framework/operator-sdk/blob/ad1d4d1953abf22c336f67198fc23a489c0f4fdb/cmd/operator-sdk/add/controller.go#L45)

hasbro17 replied:
Not in the help text of the flag itself but I can add an example in the examples section of the cmd.

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -97,60 +106,42 @@ Configure CSV generation by writing a config file 'deploy/olm-catalog/csv-config
+		crdManifestSet, err := findCRDFileSet(c.inputs[gencatalog.DeployDirKey])
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to update CRD's: %v", err)
```
May it would be failed to find?

hasbro17 replied:
Sure but that would be obvious from the actual error that's being printed out e.g `failed to read....no such file` etc.
The prefix here is to indicate something failed in the operation to update CRDs after CSV generation.

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -97,60 +106,42 @@ Configure CSV generation by writing a config file 'deploy/olm-catalog/csv-config
+		for path, b := range crdManifestSet {
+			path = filepath.Join(bundleDir, path)
+			if err = ioutil.WriteFile(path, b, fileutil.DefaultFileMode); err != nil {
+				return fmt.Errorf("failed to update CRD's: %v", err)
```
Is not the command just creating/updating CSV manifests? So, I think that this msg can bring a wrong understanding.

hasbro17 replied:
No this part of the code is after we've generated the CSV and Package Manifest files.
This part is purely to take the existing CRDs in `deploy/crds/...` and update them in the bundle directory where we generated the CSV based on the `--update-crd` flag.
So it's actually creating/updating CRDs.

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -74,14 +69,28 @@ Configure CSV generation by writing a config file 'deploy/olm-catalog/csv-config
+		map[string]string{
+			gencatalog.DeployDirKey: "deploy",
+			gencatalog.APIsDirKey:   filepath.Join("pkg", "apis"),
+		},
```
WDYT about have 2 flags instead of combining all in the same one?
Would not be more simple to do the impl and to use/understanding as well have 2 flags one for each input instead of combining them?

hasbro17 replied:
@estroz and I've discussed this before.
We want to keep the number of flags to a minimum otherwise it's going to add up and essentially become the CSV config file all over again.
That's why we're refactoring this to be a generator format with Inputs and Output.
There might be more inputs later on for CSV generation e.g CRD manifest locations.

camilamacedo86 replied:
IHMO: This approach makes the user usability and implementation very complex and it will be harder if the quantity of the options grows as you said. 

- What is the advantage/facility for the user to write/read/use `--inputs [key]=value,[key]=value` instead of `--flag=value --flag=value`?
- Also, how we will gen the doc for all possible options since it is done dynamically? 
- Note that a user expected have the full info over the command by running `--help`. How can we describe each key option and the expected value following this approach?
- If we decide to deprecate an option, how it will works? Note that users will expected check that an option is deprecated by using the helper as follows which is achieved by using  `cmd.Flags().MarkDeprecated` such as [here](https://github.com/kubernetes/kubernetes/pull/73787/files#diff-7650a51eb24776a0115b675dd3842d63R177-R178). 

> Flag --config has been deprecated, Deprecated flag config

Sorry, but I still think that is not a good idea we follow with this approach.

hasbro17 replied:
@camilamacedo86 That's a fair point actually. I think when I took over this PR, the motivation for a single `--inputs` or `--inputPaths` flag was so we could closely follow the generator pattern that controller-tools generators have.

But since then @estroz and I have agreed that the generator pattern is probably not necessary for our use case.
See: https://github.com/operator-framework/operator-sdk/pull/2511#discussion_r378385371

Since we're going to refactor the Generator config later anyway, I think removing the key/value `--inputs` flag in favor of cmd specific flags like `api-dir`, `deploy-dir` would actually be better.
So I'll make that change in this PR.

camilamacedo86 replied:
It is very cool 👍!! really tks for share the info as well.
You are terrific @hasbro17 @estroz 
Fell free to close it as resolve.

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -191,25 +182,28 @@ func validateVersion(version string) error {
+		return nil, err
+	}
+	if !info.IsDir() {
+		return nil, fmt.Errorf("crd's must be read from a directory. %s is a file", path)
```
WDYT? E.g:
```suggestion
		return nil, fmt.Errorf("The path informed by the flag `--api-dir` is for a file and not a directory")
```

hasbro17 replied:
`--api-dir` does not specify the CRD manifests location. That would be the path to the API types e.g `pkg/apis/...`.
But actually we don't even have an `--api-dir` flag.

camilamacedo86 replied:
Just to clarifies the suggestion:
> "The path informed by the flag X is for a file and not a directory as expected")

Ps. `--api-dir` is just an example since I am advocating for above we have just flags and not a flag that we can pass a keymap. However, the suggestion with the keymap would something as: 

> "The path informed by the flag [X] and in the key [Y] with the value [Z] is a path for a file and not a directory as expected")
 
is it make more sense?

camilamacedo86 commented:
Hi @aliok, 

The scenario described by you in https://github.com/operator-framework/operator-sdk/pull/2511#issuecomment-586178022  shows has no relation with this PR and added in https://github.com/operator-framework/operator-sdk/pull/1396. See [here](https://github.com/operator-framework/operator-sdk/pull/1396/files#diff-ecf668939aa4cf486d106277785b2491R143). And then, by looking in your [CSV](https://gist.github.com/aliok/47a5c393e034395e5981aeb7c1a0d979#file-knative-eventing-operator-v0-11-0-clusterserviceversion-yaml-L169) I understand that it happens because your operator has not this ENV VAR since is a non-SDK one. I think that to test this PR with you can just add it for now. 

Hi @estroz,

Shows that this scenario is a breaking change. Could we not just add  the `WATCH_NAMESPACE`  as : 
 ```
               - name: WATCH_NAMESPACE
                  valueFrom:
                    fieldRef:
                      fieldPath: metadata.namespace
```
in this scenario in order to be compatible with the non-SDK projects as well? 

c/c @hasbro17

hasbro17 commented:
> The scenario described by you in #2511 (comment) shows has no relation with this PR and added in #1396

@camilamacedo86 There's more changes incoming so please hold off further reviews at the moment. 
And @aliok's use case is pretty much what this PR is for. Turns out I forgot to account for an upgrade from an existing CSV in my unit tests and there was a subtle bug in the generator in not using the OutputDir for the expected `olm-catalog` location. 
So I've just pushed a bug fix for that but I'm adding more unit tests to cover the upgrade path from existing csv bundles in a different output dir scenario.



@aliok Thanks a lot for trying this out. Looks like I can get your csv upgraded with all the correct metadata from the previous one now.
https://gist.github.com/hasbro17/758a62f3b49de6df89cb98c7e296091b

However I still need to figure out why it's not populating your `spec.customresourcedefinitions.owned` field from your existing CRDs.
The spec and status descriptors would have to come from the [CSV annotation comments](https://github.com/operator-framework/operator-sdk/blob/master/doc/user/olm-catalog/csv-annotations.md#code-annotations-for-cluster-service-versions) defined in your `pkg/apis/...` types but I have to debug why it's not populating the bare minimum for your owned CRDs.

I'll update this once I have that sorted out.

aliok commented:
> The scenario described by you in #2511 (comment) shows has no relation with this PR and added in #1396. 

@camilamacedo86 that's from 6+ months ago and in all my trials I had that env var missing. I don't think it is really related, but I might be wrong.

@hasbro17 Just gave it a try again. Looks much better: installModes and all metadata are there but as you said no `spec.customresourcedefinitions.owned`. 

Thanks for the hard work!

hasbro17 commented:
@aliok I figured out why your apis directory was being skipped when parsing for CSV annotation comments, which results in an empty `spec.customresourcedefinitions.owned` field.

Your API types are located in `pkg/apis/eventing/v1alpha1`. However your actual group name is [`operator.knative.dev`](https://github.com/aliok/eventing-operator/blob/operator-sdk-tests-for-hasbro/pkg/apis/eventing/v1alpha1/register.go#L27).

The CSV generator will use the group name found in your CRD manifest to look for the associated type definition in `pkg/apis/operator/v1alpha1`. So that's why it was skipping when it couldn't find the types in the expected directory.

However I think we should be able to relax this restriction for single group APIs so that you can have `some/random/path/...` for your api types directory. That's your current case.
But for multi-group apis I think the CSV generator might require `<api-dir>/<group>/<version>`. I'll have to double check this with @estroz

hasbro17 commented on internal/generate/olm-catalog/csv_updaters.go:
```diff
@@ -229,20 +231,64 @@ var _ csvUpdater = crds{}
 // apply will only make a new spec.customresourcedefinitions.owned element for
 // a type if an annotation is present on that type's declaration.
```
@estroz I don't think this is true right?
I can have no CSV annotations defined on my types and the CSV generator will still generate the bare minimum `spec.customresourcedefintions.owned` with my gvks populated.

estroz replied:
Yes this comment should have been removed in a preceding PR.

hasbro17 commented:
@estroz @camilamacedo86 After lots of misadventures with gengo and much hand wringing over how to sidestep them, this should be ready for another review now.

Noticeable updates:
- Refactored `internal/generate/olm-catalog/descriptor/descriptor.go` quite a fair bit to allow the CSV generator to parse api directories with different layouts. Previously it was hardcoded to only look at `<api-dir>/<group>/<version>`. I've expanded this to add `<api-dir>/<version>`  and to recursively look for `<version>` as well. We'll need this for kubebuilder's multi and single group layouts. @estroz PTAL at this in particular.
- Added quite a few unit tests for the descriptor and csv tests to cover different cases for multi/single group pkg layouts, different input/outputs and upgrades from existing CSVs. You might want to ignore the additional test manifests.
- Replaced the key/value `--inputs` flags with `--deploy-dir --apis-dir --crd-dir`. @camilamacedo86 This was per our earlier discussion so we can better maintain individual flags.
- A subtle bug where the CSV generator did not error out when it failed to walk the deploy directory for manifests. Fixed and added a specific test for this.
- TODOs and caveats marked for later.

hasbro17 commented:
@aliok If you still have time to give this another shot I think I've handled all the issues from before.
Note that the `--input` flag has been replaced with `--deploy-dir --apis-dir` for specifying the api and deploy directories. See the [CLI reference](https://github.com/operator-framework/operator-sdk/blob/5027f7241a6b260d69c1b7d0815e39424bd277a1/doc/cli/operator-sdk_generate_csv.md#options) for that update.

Based on the branch you've provided earlier the following should work. 
```sh
operator-sdk generate csv \
--csv-version="0.12.0" \
--from-version="0.11.0" \
--deploy-dir=config \
--operator-name=knative-eventing-operator \
--csv-channel=alpha \
--default-channel \
--update-crds
```

Although as I mentioned before, your existing metadata in `spec.customresourcedefintions.owned` such as spec/status descriptors won't be generated since you need to add the [CSV annotation comments](https://github.com/operator-framework/operator-sdk/blob/master/doc/user/olm-catalog/csv-annotations.md#code-annotations-for-cluster-service-versions) in your type definitions for that.


Lastly while the CSV generator should work for your case, my earlier comment on the layout of the API types directory still holds:

>Your API types are located in pkg/apis/eventing/v1alpha1. However your actual group name is operator.knative.dev.
>
>The CSV generator will use the group name found in your CRD manifest to look for the associated type definition in pkg/apis/operator/v1alpha1. So that's why it was skipping when it couldn't find the types in the expected directory.

The CSV generator normally expects a layout of `<api-dir>/<group>/<version>` for multiple groups or `<api-dir>/<version>` for single group APIs.
The layout in the knative eventing-operator is neither since it has the wrong group name in the layout.
So the CSV generator makes a best effort to find the right API types package for `v1alpha1` but this will be prone to collisions if you add multiple groups which also have the same version.
So it's okay for now but ideally you might want to correct the pkg layout in the future if possible.

hasbro17 commented:
@camilamacedo86 The linter in CI [is complaining](https://travis-ci.org/operator-framework/operator-sdk/jobs/659048186#L899) about maligned structs:
```console
$ golangci-lint run --disable-all --enable=maligned
internal/generate/olm-catalog/descriptor/descriptor_test.go:43:16: struct of size 120 bytes could be of size 112 bytes (maligned)
	subTests := []struct {
```
Have you seen this before and how to fix this when you first turned on this flag?

jmrodri commented on internal/generate/olm-catalog/descriptor/descriptor_test.go:
```diff
@@ -26,89 +26,154 @@ import (
-		numPkgTypes int
-		wantNil     bool
+		// True if apis dir in the expected single or multi group layout
+		isExpectedLayout bool
```
@hasbro17 move these 2 lines below the `wantNil bool`. This will fix the `maligned` error. https://paste.centos.org/view/017c27be
Having the `bool` between all of the `string`s causes a maligned struct in memory. https://medium.com/@sebassegros/golang-dealing-with-maligned-structs-9b77bacf4b97

Basically with your current layout it would look like this (X = 1 byte for the variable, P = padding)
```
string  16 XXXXXXXX
           XXXXXXXX
bool     1 X 
            PPPPPPP
string  16 XXXXXXXX
           XXXXXXXX
string  16 XXXXXXXX
           XXXXXXXX
string  16 XXXXXXXX
           XXXXXXXX
string  16 XXXXXXXX
           XXXXXXXX
string  16 XXXXXXXX
           XXXXXXXX
int      4 XXXX
bool     1     X   
                PPP
```
By moving the bool to the bottom of the struct you get only 2 bytes of padding vs 10.
```
string  16 XXXXXXXX
           XXXXXXXX
string  16 XXXXXXXX
           XXXXXXXX
string  16 XXXXXXXX
           XXXXXXXX
string  16 XXXXXXXX
           XXXXXXXX
string  16 XXXXXXXX
           XXXXXXXX
string  16 XXXXXXXX
           XXXXXXXX
int      4 XXXX
bool     1     X   
bool     1      X   
                 PP
```
I did this locally and the maligned error went away.

hasbro17 replied:
@jmrodri Fascinating. Thanks for the fix and explanation.

estroz commented on internal/generate/olm-catalog/descriptor/descriptor.go:
```diff
@@ -119,16 +150,57 @@ func GetCRDDescriptionForGVK(apisDir string, gvk schema.GroupVersionKind) (olmap
 
-// getTypesFromDir gets all Go types from dir.
-func getTypesFromDir(dir string) (types.Universe, error) {
+func isDirExist(path string) (bool, error) {
```
The semantics of this function are weird, since it returns an error if the dir exists but has some stat error. Its callers should be expected to handle the case of the dir existing but not being accessible.

hasbro17 replied:
> Its callers should be expected to handle the case of the dir existing but not being accessible

@estroz That's exactly the only case when an error is returned here i.e stat error when file exists but inaccessible (permission or filesystem) error.
```Go
if err != nil {
	if os.IsNotExist(err) {
		return false, nil
	}
	return false, err
}
```
So not sure what you wanted to change about the error return for this function.

All the callers of this function are currently "handling" this "exists-but-inaccessible" error by throwing it all the way to the top causing `generate csv` to exit with an error. 
Although on double checking, one caller in `GetCRDDescriptionForGVK()` was ignoring this error but I've fixed that.

estroz replied:
Given the name, I would expect something like:

```go
func isDirExist(path string) bool {
	info, err := os.Stat(path)
	if err != nil {
		return os.IsExist(err)
	}
	return info.IsDir()
}
```

Considering how you're using this function, a better name would be `isDirAccessible`.

estroz commented on internal/generate/olm-catalog/descriptor/descriptor.go:
```diff
@@ -51,19 +51,50 @@ func GetCRDDescriptionForGVK(apisDir string, gvk schema.GroupVersionKind) (olmap
+		// TODO: gengo.parse.AddDirRecursive() will (sometimes?) fail if the
+		// root apisDir has no .go files.
+		// Workaround for this is to have a doc.go file in the package.
+		// Move away from using gengo in the future if possible.
```
Yes I'd very much like to move to controller-tools' parser implementation instead.

estroz commented on internal/generate/olm-catalog/descriptor/descriptor.go:
```diff
@@ -119,16 +150,57 @@ func GetCRDDescriptionForGVK(apisDir string, gvk schema.GroupVersionKind) (olmap
+
+	// Gengo's AddDirRecursive fails to load subdir pkgs if the root dir
+	// isn't the full pkg import path.
+	dirImportPath := path.Join(projutil.GetGoPkg(), filepath.ToSlash(dir))
```
What is wrong with using `dir` directly?

hasbro17 replied:
Can't use `dir` directly since gengo complains it doesn't begin with `./`.
So did a bit of testing and seems like I can use `./<dir>` for AddDirRecursive() and it will work fine.
I've made the changes [here](https://github.com/operator-framework/operator-sdk/pull/2511/commits/655314e6864cc7554f833bd22b70eaa2dc350956#diff-56681196d4230f6aa4716b9c2ee4851e) and that's simplified a lot filepaths in the unit tests as well. Plus now we don't have to rely on constructing the full pkg import paths.

estroz commented on internal/generate/olm-catalog/package_manifest.go:
```diff
@@ -60,21 +63,41 @@ func NewPackageManifest(cfg gen.Config, csvVersion, channel string, isDefault bo
 	return g
 }
 
+func isFileExist(path string) bool {
```
Similar comment to that for `isDirExist`.

estroz replied:
If we're keeping this function, better to return an error and rename to `isFileAccessible` so we don't exit the program.

estroz commented on internal/generate/olm-catalog/package_manifest.go:
```diff
@@ -60,21 +63,41 @@ func NewPackageManifest(cfg gen.Config, csvVersion, channel string, isDefault bo
+	pkgManifestDirPath := filepath.Join(olmCatalogDir, g.OperatorName)
+	pkgManifestFilePath := filepath.Join(pkgManifestDirPath, g.fileName)
+	if isFileExist(pkgManifestFilePath) {
+		g.existingPkgManifestDir = pkgManifestDirPath
```
If this is only being used once in `buildPackageManifest`, it doesn't need to be a field and should instead be constructed in that method.

aliok commented:
@hasbro17 

I think we're in a very very good shape. I now can use `operator-sdk` to generate CSV for upstream Knative eventing operator.

Thank you very much for your hard work!

> Although as I mentioned before, your existing metadata in spec.customresourcedefintions.owned such as spec/status descriptors won't be generated since you need to add the CSV annotation comments in your type definitions for that.

That's fine. Nothing you need to worry about. Thanks for the heads up.

> Your API types are located in pkg/apis/eventing/v1alpha1. However your actual group name is operator.knative.dev.

Same as above. Thanks!

From my point of view, we can merge this PR and I can start using it. Thanks very much!

hasbro17 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -53,35 +55,58 @@ for the operator. This file is used to publish the operator to the OLM Catalog.
 			}
+			// Default for crd dir if unset
+			if c.crdDir == "" {
+				c.crdDir = c.deployDir
```
It would still be necessary to change the signature.
Users can set `--crd-dir` to whatever they want.
So if they set it to `--crd-dir=deploy` or some other directory which has the `olm-catalog` as a subdirectory then `GetCRDManifestPaths()` needs to ignore that in the filepath Walk.

The signature change is mostly because the `GetCRDManifestPaths()` util is shared between `generate crd` where the crd directory is always fixed to `deploy/crd` and `generate csv --crd-dir` where the CRD directory could be any path.
I could split this into two separate almost identical utils for each command but this is a smaller change.

Also `--crd-dir=deploy` or `c.crdDir = c.deployDir` because that's essentially what it was before this refactor when there was no `crd-dir` flag. Let me know if you think we should change that default though, unrelated to the `GetCRDManifestPaths()` signature change.

estroz replied:
In that case, we should:

1. Change `GetCRDManifestPaths` so that it does a shallow (depth 1) search for CRDs. This is more intuitive IMO, since there's nothing in the name that says it performs a recursive search. Now when users pass a dir without CRDs in it (ex. `deploy`) they should see an error that no CRDs were present in that dir.
1. Pass `crdDir` to `Config.Inputs` for the CRD generator, which should be looking in that directory anyway.
1. Default `crdDir` to `filepath.Join(c.deployDir, "crds")`.

Where is `GetCRDManifestPaths` used by `generate crds`?

hasbro17 replied:
@estroz `GetCRDManifestPaths()` is used by `generate crds` here:
https://github.com/operator-framework/operator-sdk/blob/master/internal/util/k8sutil/crd.go#L32

If we're doing a shallow read instead I'll leave that util untouched and just do a simple read in `generate csv`.

estroz replied:
Remove this default since it is now set during flag registration.

hasbro17 commented:
@estroz I've addressed your comments above. Plus as I was refactoring I realized one case where the owned CRD description updates were being silently skipped if the project type wasn't determined as Go type for some reason (e.g different layout).

I've refactored `updateDescriptions()` to make that case more clear and add a log for when that happens. You can see the [commit](https://github.com/operator-framework/operator-sdk/pull/2511/commits/a2f16f852e0aadfbb4361385523a75882b4c1b07) just  for that.

hasbro17 commented:
@camilamacedo86 PTAL as well. The csv generate cmd should now have separate flags for its inputs per our [earlier discussion](https://github.com/operator-framework/operator-sdk/pull/2511#discussion_r379667457).

camilamacedo86 commented on internal/util/projutil/project_util.go:
```diff
@@ -113,9 +113,19 @@ func getHomeDir() (string, error) {
 	return homedir.Expand(hd)
 }
 
+// TODO(hasbro17): If this function is called in the subdir of
```
Should be here the best place for it? WDYT about track an issue as RFE for that with all info as to how to reproduce the scenario and the goal? Then, we can try to analyse it and see if it is possible or not?

hasbro17 replied:
@camilamacedo86 I'm planning to compile the entire list of TODOs here into a follow up ticket/issue for more general refactoring work. 
If I stop and file an issue for every TODO I come across this PR will never end, so I'm trying to keep the scope minimal for now.

That being said, this particular caveat is not an issue right now. So there's no "issue" to reproduce.
It's something that could be an issue if this util is used with the wrong expectation. Which is why I've only noted this down with the expectation of addressing this later if that's even possible.

camilamacedo86 commented on internal/generate/olm-catalog/csv_updaters.go:
```diff
@@ -226,39 +228,86 @@ var _ csvUpdater = crds{}
+				log.Infof("No kind type found for API %s. Skipping CSV annotation parsing for API.", gvk)
+			} else {
+				// TODO: Should we ignore all CSV annotation parsing errors and simply log the error
+				// like we do for the above cases.
```
```suggestion
// CSV annotation parsing errors will throw error. 
```
I think we could make a comment  here to make clear the current behaviour but the question could `should  we throw an error or just output warnings instead?` be raised in an issue. Then, IMO in this task we should first at all check what is the default behaviour in this scenario with kubebuilder/controller-tools and follow up it. WDYT?

hasbro17 replied:
As mentioned above I'm going to file issues for these TODOs but I still want to keep them here so I can point to them later.
I think the line below `return fmt.Errorf("failed to set CRD descriptors for %s: %v", gvk, err)` makes it clear that we're returning an error so I don't think I need to add a comment here for that.

camilamacedo86 commented on internal/generate/olm-catalog/csv_updaters.go:
```diff
@@ -226,39 +228,86 @@ var _ csvUpdater = crds{}
+func updateDescriptions(csv *olmapiv1alpha1.ClusterServiceVersion, searchDir string) error {
+	// TODO: Should this skip generating descriptions if API directory exists
+	// but project type can't be detected due to layout differences.
+	if opType := projutil.GetOperatorType(); opType != projutil.OperatorTypeGo {
```
Currently is not supportable but is possible gen the CSV for Helm/Ansible as I understand that the goal is we support it as well. So,  I do not think that we should have this because of this and because we do not do this skip for non-SDK projects as well. In this way, why to do it for helm/ansible types?

hasbro17 replied:
@camilamacedo86 You can test this out locally but this does not skip generating CSVs for Helm/Ansible types. This only skips generating the owned CRD descriptions in the CSV. 
This kept the existing behavior for `generate csv` where it will skip `updateDescriptions()` and won't parse `pkg/apis` for CSV annotations for helm/ansible types since they don't have CSV annotations yet.

But your comment did make me reconsider why I'm explicitly detecting the Go project type when all we care about is parsing CSV annotations from the API directory e.g `pkg/apis`. 
Since we already handle the case of skipping annotation parsing for that case below I''m going to remove this check.

hasbro17 replied:
Removed this check: https://github.com/operator-framework/operator-sdk/pull/2511/commits/3b3d3e08f4935a0c63524e9e7dc510980e10ef3d

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -96,61 +121,49 @@ Configure CSV generation by writing a config file 'deploy/olm-catalog/csv-config
+	cfg := gen.Config{
+		OperatorName: c.operatorName,
+		// TODO(hasbro17): Remove the Input key map when the Generator input keys
+		// are removed in favour of config fields in the csvGenerator
```
The comment here is not very clear for me. What is the requirement for we are able to remove it? Can we link an issue to know what is expected to get done first? Could we clarify more this? WDYT?

hasbro17 replied:
> Can we link an issue to know what is expected to get done first?

@camilamacedo86 This is the follow up work that needs to happen after this PR gets merged. @estroz and I discussed this above:
https://github.com/operator-framework/operator-sdk/pull/2511#discussion_r378385371

As mentioned above I will file the actual issue(s) after this is merged because it only becomes an issue after this is merged.

**In summary:** We will replace the `Inputs` key value map with properly defined fields in the CSV and CRD generator structs.
The reason is to to reduce the complexity of passing inputs to the CSV generator so it's easier to understand and maintain.

Quite similar to how we replaced the key value inputs flag `inputs=<key1>=<val1>,<key2>=<val2>` flag with defined flags like `--deploy-dir=deploy`, `--apis-dir=pkg/apis` etc. I mentioned this in response to your earlier review comments as well.
https://github.com/operator-framework/operator-sdk/pull/2511#discussion_r380383154

Again I'm keeping it out of this PR to keep the scope minimal for something that will be a purely internal change with no user facing or CLI changes.

camilamacedo86 commented on doc/cli/operator-sdk_generate_csv.md:
````diff
@@ -11,22 +11,42 @@ A CSV semantic version is supplied via the --csv-version flag. If your operator
+
+```
+
+TODO
````
?

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -53,35 +55,58 @@ for the operator. This file is used to publish the operator to the OLM Catalog.
+Note: The CSV generator only uses this to copy the CRD manifests.
+The CSV contents for spec.customresourcedefinitions.owned will still be updated
+from the CRD manifests in the deploy directory specified by --deploy-dir.
+If unset, it defaults to the same value as --deploy-dir.
```
I think the above text is too verbose. Just `Project relative path to the root directory for CRD manifests` shows enough here. Could you w not add the rest in so example?

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -53,35 +55,58 @@ for the operator. This file is used to publish the operator to the OLM Catalog.
+fill out metadata for owned APIs in spec.customresourcedefinitions.owned.
+`)
+	cmd.Flags().StringVar(&c.crdDir, "crd-dir", "",
+		`Project relative path to root directory for for CRD manifests.
```
```suggestion
		`CRD manifests path dir directory (allow relative path from the root project)`. 
```

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -53,35 +55,58 @@ for the operator. This file is used to publish the operator to the OLM Catalog.
+	cmd.Flags().StringVar(&c.apisDir, "apis-dir", filepath.Join("pkg", "apis"),
+		`Project relative path to root directory for API type defintions.
+The CSV annotation comments will be parsed from the Go types under this path to
+fill out metadata for owned APIs in spec.customresourcedefinitions.owned.
```
```suggestion
		`API type definitions directory path (allows relative paths from to root project directory).
```
Is not better be less verbose and the rest of the info be added as a note in an example of its usage here?

camilamacedo86 replied:
See: https://github.com/camilamacedo86/operator-sdk/blob/master/cmd/operator-sdk/add/controller.go#L44

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -53,35 +55,58 @@ for the operator. This file is used to publish the operator to the OLM Catalog.
 		RunE: func(cmd *cobra.Command, args []string) error {
-			// The CSV generator assumes that the deploy and pkg directories are
-			// present at runtime, so this command must be run in a project's root.
-			projutil.MustInProjectRoot()
```
HI @hasbro17, 

If we execute the command from another dir which is not the root of the project it shows that worked but actually not. I know that the goal is to allow it to work with  no-sdk project as well. However, IMO we should check and add an WARN. 

```
        if err := CheckProjectRoot(); err != nil {
		// log warn such as `WARN[0000] Ensure that you are running this command from the project root dir. `
	}
```

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -53,35 +55,58 @@ for the operator. This file is used to publish the operator to the OLM Catalog.
-		"Path to CSV config file. Defaults to deploy/olm-catalog/csv-config.yaml")
+
+	cmd.Flags().StringVar(&c.deployDir, "deploy-dir", "deploy",
+		`Project relative path to root directory for operator manifests (Deployment, RBAC, CRDs).
```
Allows relative and full path. So, wdyt about something such as `Operator Manifest directory path where are the Deployment, RBAC and CRDs. (E.g deploy/) (allows relative paths from to root project directory).`

Note that thee root word his shows not necessary.

camilamacedo86 commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -53,35 +55,58 @@ for the operator. This file is used to publish the operator to the OLM Catalog.
+`)
+
+	cmd.Flags().StringVar(&c.outputDir, "output-dir", scaffold.DeployDir,
+		"Base directory to output generated CSV. The resulting CSV bundle directory"+
```
```suggestion
	The path for the directory where the CSV will be generated. The resulting CSV bundle directory"+
```

aliok commented:
@hasbro17 let me know if you need another test drive when you make changes. I will be happy doing that

estroz commented on cmd/operator-sdk/generate/csv.go:
```diff
@@ -53,35 +55,58 @@ for the operator. This file is used to publish the operator to the OLM Catalog.
-version to --from-version. Otherwise the SDK will scaffold a new CSV manifest.
+version to --from-version. Otherwise the SDK will scaffold a new CSV manifest.`,
+		Example: `
+TODO
```
Example needs to be done before merge.

estroz commented on internal/generate/olm-catalog/descriptor/descriptor.go:
```diff
@@ -51,19 +49,55 @@ func GetCRDDescriptionForGVK(apisDir string, gvk schema.GroupVersionKind) (olmap
+		return olmapiv1alpha1.CRDDescription{}, ErrAPIDirNotExist
+	}
+
+	// Check if the kind pkg is at the expected layout
```
Is there a reason to check for different layouts? Given `apisDir` is configurable, `getTypesFromDirRecursive()` will find types regardless right?

hasbro17 replied:
So `apisDir` is only configurable to be the API types root dir from where we construct an exact path to the types subdirectory for a GVK e.g `<api-dir>/<group>/<version>` or `<api-dir>/<version>`.

For muli-group layouts e.g `pkg/apis/cache/v1alpha1` and `pkg/apis/app/v1alpha1` if we just parse `pkg/apis` recursively and then try to find the types for the GVK `app/v1alpha1 App` from the universe of pkgs we'll have a collision when looking for a pkg named `v1alpha1`.

So that's why we first detect the exact layout for multi or single group pkgs and only parse the exact pkg dir and not the whole of `pkg/apis`.

Also I've realized `getTypesFromDirRecursive()` is named incorrectly since it returns the universe of pkgs [`type Universe map[string]*Package`](https://github.com/kubernetes/gengo/blob/master/types/types.go#L224) from the dir. The exact pkg types `[]*types.Type` are then filtered out by `getTypesForPkgPath`.

Going to rename `getTypesFromDirRecursive()` to `getPkgsFromDirRecursive()`

estroz commented on internal/util/projutil/project_util.go:
```diff
@@ -180,6 +190,11 @@ func GetOperatorType() OperatorType {
 	return OperatorTypeUnknown
 }
 
+func hasGoPkgFile() bool {
```
Is this being used anywhere?

hasbro17 commented:
@estroz @camilamacedo86 PTAL again.

I had to rebase again to fix the CHANGELOG but you can take  a look at just the [last commit](https://github.com/operator-framework/operator-sdk/pull/2511/commits/6ef5a74cbb7b5b9e820a699499c0a8407cb447f5).
I've added the cobra cmd example for default and custom inputs paths.

@camilamacedo86 I've reduced the flag help text verbosity by moving all the extra caveats and info for each flag to the `Long` section.
Also added a warning if the command is not running from project root. Not error-ing out like we did before because not all projects have "build/Dockerfile" at the root.

@aliok Thanks for offering to test this again but I think this should be alright. It's mostly the same from when you last tested it. Just added the finishing touches for docs etc and we should hopefully have this merged very soon.

hasbro17 commented:
Missed this comment: https://github.com/operator-framework/operator-sdk/pull/2511/files#r396573444
Going to address the issue of `--crd-dir` flag only being used to copy CRD manifests but not being used to source those manifests for when we actually populate the CSV's `spec.customresourcedefinitions.owned`.

hasbro17 commented:
@estroz PTAL.
I've removed the overlap between `--deploy-dir` and `--crd-dir` so that now `updateCSVFromManifests()` will use the path from `--crd-dir` to parse the CRD CR manifests.

And we're no longer doing recursive `filepath.Walk()`s for either deploy or crds directories.

**Side-note:** This brings up another issue for me on the kubebuilder integration side where CRD and CR manifests are separate directories, and so are Deployment and RBAC manifests.
But I'm leaving that for later. 
I've added TODO's to allow multiple paths for `--deploy-dir` and `--crd-dir` in the future.

estroz commented:
@hasbro17 to fix the unit test error:

```sh
$ mkdir -p ./internal/generate/testdata/go/emptydir
$ touch ./internal/generate/testdata/go/emptydir/.keep
```

I had to do this in #2729.

estroz commented on internal/generate/olm-catalog/csv.go:
```diff
@@ -15,15 +15,490 @@
+	}
+
+	if crdsDir, ok := g.Inputs[CRDsDirKey]; !ok || crdsDir == "" {
+		g.Inputs[CRDsDirKey] = filepath.Join(scaffold.DeployDir, scaffold.CRDsDir)
```
```suggestion
		g.Inputs[CRDsDirKey] = filepath.Join(g.Inputs[DeployDirKey], "crds")
```

hasbro17 commented:
*Sigh* 
@estroz Any clue why the scorecard test for bundle validation might be failing? I'm not sure if after rebasing from master if the scorecard now relies on any testdata manifests that I've modified/moved around in this PR.
https://travis-ci.org/github/operator-framework/operator-sdk/jobs/669505119#L1244
```console
olm:
	Bundle Validation Test              
	CR: example-memcached
	Labels: 
		"necessity":"required"
		"suite":"olm"
		"test":"bundlevalidationtest"
	Errors: 
		unable to find the OLM '\''bundle'\'' directory which is required for this test
	Provided APIs have validation       pass
	CR: example-memcached
	Labels: 
		"test":"crdshavevalidationtest"
		"necessity":"required"
		"suite":"olm"
```
Or if scorecard is using generate csv to generate the bundle directory and somehow failing to.
Going to dig into this a bit.

estroz commented on test/test-framework/deploy/olm-catalog/memcached-operator/0.0.2/memcached-operator.v0.0.2.clusterserviceversion.yaml:
```diff
@@ -1,30 +1,109 @@
+        path: reason
+        x-descriptors:
+        - urn:alm:descriptor:io.kubernetes.phase:reason
+      version: v1alpha1
```
```suggestion
      version: v1alpha1
    - kind: NotExistKind
      name: notexist.example.com
      version: v1alpha1

```
should fix the scorecard error

hasbro17 commented:
@estroz Thanks for the fix 🙏 
Had to reset CI a few times to get marker to play nice but all clear in the end.
